questionpy_server/cache.py

```diff
@@ -0,0 +1,160 @@
+
+        directory = Path(path).resolve()
+        directory.mkdir(exist_ok=True)
+
+        self._directory: Path = directory
+        self._max_size: int = size
+        self._extension: str = '' if extension is None else '.' + extension.lstrip('.')
+
+    @property
+    def max_size(self) -> int:
+        return self._max_size
+
+    @property
+    def directory(self) -> Path:
+        # Ensures that the cache directory exists.
+        if not self._directory.is_dir():
+            self._directory.mkdir(parents=True, exist_ok=True)
+        return self._directory
+
+    def _path(self, filename: str) -> Path:
+        """
+        Creates a path of the given filename. It is not mandatory that the filename exists.
+
+        :param filename: the filename
+        :return: path of the file
+        """
+
+        return self.directory / (filename + self._extension)
+
+    def exists(self, filename: str) -> bool:
+        """
+        Checks if the file is in cache.
+
+        :param filename: the filename
+        :return: True if file exists, else False
+        """
+
+        return self._path(filename).is_file()
+
+    def get(self, filename: str) -> Path:
+        """
+        Retrieves the path of a file in cache.
+
+        :param filename: the filename
+        :return: path of the file
+        """
+
+        path = self._path(filename)
+
+        if path.is_file():
+            return path
+
+        raise FileNotFoundError
+
+    def size(self) -> int:
+        """
+        Calculates cache size in bytes.
+
+        :return: cache size in bytes.
+        """
+
+        files = self.files()
+        stats = files.values()
+        return sum(stat.st_size for stat in stats)
+
+    def files(self) -> Dict[Path, stat_result]:
+        """
+        Creates a dictionary consisting of the path of every file in cache and their corresponding stats.
+
+        :return: paths and stats of every file in cache
+        """
+
+        file_stats = {}
+
+        for entry in self.directory.iterdir():
+            if entry.is_file():
+                file_stats[entry] = entry.stat()
+
+        return file_stats
+
+    def clear(self) -> None:
+        """
+        Removes every file inside the cache directory.
+        """
+
+        files = self.files()
+        for file in files:
+            file.unlink(missing_ok=True)
+
+    def free(self, amount: int) -> int:
+        """
+        Removes oldest accessed files until a minimum amount of cache storage is freed.
+
+        :param amount: minimum amount of cache to be freed
+        :return: real number of bytes freed
+        """
+
+        if amount > self._max_size:
+            raise CacheTooSmallError(f'Can not free {amount} bytes as the cache can only hold {self._max_size} bytes.')
+
+        # Get files and sort them by access time.
+        file_stats = self.files()
+        file_stats_sorted = sorted(file_stats.items(), key=lambda item: item[1].st_atime_ns)
```

Comment:
Das ist zu sehr vom Dateisystem abhängig (oft wird mit `noatime` gemountet). Ich würde die Cache-Keys (also die Dateinamen) bei Programmstart in einen `OrderedDict` einlesen und aktuell halten. Beim Zugriff auf einen Inhalt wird `move_to_end` aufgerufen.
Wenn direkt nach dem Programmstart Elemente aus dem Cache entfernt werden müssen, arbeitet der Cache zwar zunächst nicht nach LRU, aber das ist vertretbar.